File: Level 3/순위/solution.py
from collections import defaultdict


def solution(n, results):
    winners = defaultdict(set)
    losers = defaultdict(set)

    for winner, loser in results:
        losers[winner].add(loser)
        winners[loser].add(winner)

    for i in range(1, n + 1):
        for winner in winners[i]:
            losers[winner].update(losers[i])
        for loser in losers[i]:
            winners[loser].update(winners[i])

    return sum(1 for i in range(1, n + 1) if len(winners[i]) + len(losers[i]) == n - 1)

File: Level 3/순위/solution.js
function solution(n, results) {
  const winners = Array.from({ length: n + 1 }, () => new Set());
  const losers = Array.from({ length: n + 1 }, () => new Set());

  results.forEach(([winner, loser]) => {
    winners[loser].add(winner);
    losers[winner].add(loser);
  });

  for (let i = 1; i <= n; i++) {
    winners[i].forEach((winner) => {
      losers[i].forEach((loser) => losers[winner].add(loser));
    });
    losers[i].forEach((loser) => {
      winners[i].forEach((winner) => winners[loser].add(winner));
    });
  }

  let answer = 0;

  for (let i = 1; i <= n; i++) {
    if (winners[i].size + losers[i].size === n - 1) {
      answer++;
    }
  }

  return answer;
}
